presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1507,6 +1509,161 @@ private static void validateSphericalType(String function, OGCGeometry geometry,
+
+        for (int i = start; i < end; i++) {
+            polygon.getPoint(i, tempOutPoint);
+            calculator.processPoint(tempOutPoint);
+        }
+
+        return calculator.getSphericalExcess();
+    }
+
+    private static class BearingAndExcessCalculator
+    {
+        private static final double TWO_PI = 2 * Math.PI;
+        private static final double THREE_PI = 3 * Math.PI;
+
+        private double sphericalExcess;
+        private double courseDelta;
+
+        private boolean hasProcessedPoint;
+        private double firstInitialBearing;
+        private double previousBearing;
+
+        private double phiPreviousPoint;
+        private double cosPreviousPoint;
+        private double sinPreviousPoint;
+        private double tanPreviousPoint;
+        private double longitudePreviousPoint;
+
+        public BearingAndExcessCalculator(Point endPoint)
+        {
+            phiPreviousPoint = toRadians(endPoint.getY());
+            sinPreviousPoint = Math.sin(phiPreviousPoint);
+            cosPreviousPoint = Math.cos(phiPreviousPoint);
+            tanPreviousPoint = Math.tan(phiPreviousPoint / 2);
+            longitudePreviousPoint = toRadians(endPoint.getX());
+        }
+
+        private void processPoint(Point p)
+        {
+            double phiCurrentPoint = toRadians(p.getY());
+            double tanCurrentPoint = Math.tan(phiCurrentPoint / 2);
+            double longitudeCurrentPoint = toRadians(p.getX());
+
+            // Only process points which are different from the previous one,
+            // Otherwise calculating the bearing is not deterministic
+            if (longitudeCurrentPoint != longitudePreviousPoint || phiCurrentPoint != phiPreviousPoint) {
+                double deltaLambda = longitudeCurrentPoint - longitudePreviousPoint;
+                sphericalExcess += 2 * Math.atan2(Math.tan(deltaLambda / 2) * (tanPreviousPoint + tanCurrentPoint), 1 + tanPreviousPoint * tanCurrentPoint);
+
+                double cosCurrentPoint = Math.cos(phiCurrentPoint);
+                double sinCurrentPoint = Math.sin(phiCurrentPoint);
+                double sinDeltaLambda = Math.sin(deltaLambda);
+                double cosDeltaLambda = Math.cos(deltaLambda);
+
+                // Initial bearing from previous to current
+                double y = sinDeltaLambda * cosCurrentPoint;
+                double x = cosPreviousPoint * sinCurrentPoint - sinPreviousPoint * cosCurrentPoint * cosDeltaLambda;
+                double initialBearing = (Math.atan2(y, x) + TWO_PI) % TWO_PI;
+
+                // Final bearing from previous to current = opposite of bearing from current to previous
+                double finalY = -sinDeltaLambda * cosPreviousPoint;
+                double finalX = sinPreviousPoint * cosCurrentPoint - cosPreviousPoint * sinCurrentPoint * cosDeltaLambda;
+                double finalBearing = (Math.atan2(finalY, finalX) + PI) % TWO_PI;
+
+                if (hasProcessedPoint) {
```

Comment:
Perhaps, initialize courseDelta to -1 and use `if (courseDelta != -1)` instead of using extra flag.

Reply:
I'm wondering if the value could get negative during the calculation ... probably on some complex polygons

Reply:
actually, it can totally go negative, as a counter clockwise polygon will end up at -360